### PR TITLE
[REVIEW] Add DecimalDtype to cuDF

### PR DESCRIPTION
**Pull request by codereport**
This resolves https://github.com/rapidsai/cudf/issues/6654.

Will update description later.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by shwina** [comment]
Assigned @kkraus14 to review as I can't review a PR I'm participating in :)

File: python/cudf/cudf/core/dtypes.py [hunk 1]
```diff
@@ -211,3 +212,32 @@ def __eq__(self, other):
 
     def __repr__(self):
         return f"StructDtype({self.fields})"
+
+
+class DecimalDtype(ExtensionDtype):
+
+    name = "decimal"
+    _metadata = ("precision", "scale")
```
Comment (kkraus14) [inline python/cudf/cudf/core/dtypes.py:220]: What is this `_metadata` attribute used for?

File: python/cudf/cudf/core/dtypes.py [hunk 2]
```diff
@@ -211,3 +212,32 @@ def __eq__(self, other):
...
+        return self._typ.precision
+
+    @property
+    def scale(self):
+        return self._typ.scale
+
+    @property
+    def type(self):
+        # might need to account for precision and scale here
+        return decimal.Decimal
```
Comment (kkraus14) [inline python/cudf/cudf/core/dtypes.py:236]: I believe the python `decimal` library doesn't allow setting the precision / scale when constructing objects as it does both fixed point and floating point decimal math. Maybe we should consider an alternative object here?

File: python/cudf/cudf/core/dtypes.py (hunk 1 above)
Comment (shwina) [reply python/cudf/cudf/core/dtypes.py:220]: It's documented [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.api.extensions.ExtensionDtype.html#pandas.api.extensions.ExtensionDtype).

Essentially, if you have an ExtensionDtype that encapsulates a bunch of attributes, you should list those in the `_metadata` attribute. Then you get things like `__eq__` for free: https://github.com/pandas-dev/pandas/blob/v1.1.4/pandas/core/dtypes/base.py#L96-L124

File: python/cudf/cudf/core/dtypes.py (hunk 2 above)
Comment (shwina) [reply python/cudf/cudf/core/dtypes.py:236]: Yup -- open to suggestions here. `.type` seems to be a question mark for all our extension dtypes...

File: python/cudf/cudf/core/dtypes.py (hunk 1 above)
Comment (kkraus14) [reply python/cudf/cudf/core/dtypes.py:220]: Thanks!

File: python/cudf/cudf/core/dtypes.py (hunk 2 above)
Comment (kkraus14) [reply python/cudf/cudf/core/dtypes.py:236]: Agreed. We can leave this for now I suppose, but we really should have some type of class here that allows us to construct `cudf.Scalar` objects in the future.

Comment (shwina) [reply python/cudf/cudf/core/dtypes.py:236]: Just `cudf.Scalar`, I suppose?

Comment (kkraus14) [reply python/cudf/cudf/core/dtypes.py:236]: The fear in doing that is I believe there's a lot of code out there that uses `.type` expecting distinct classes for different Scalars. I.E. this is in PyArrow: https://github.com/apache/arrow/blob/f40507545c9c303e85799a874833a4ab48556071/python/pyarrow/pandas_compat.py#L83-L103

Comment (shwina) [reply python/cudf/cudf/core/dtypes.py:236]: Makes sense. Punt for now?

Comment (kkraus14) [reply python/cudf/cudf/core/dtypes.py:236]: Yea I suppose. I think the answer is we need classes that wrap our new Scalar class that are strongly typed.

Comment (shwina) [reply python/cudf/cudf/core/dtypes.py:236]: Yeah could be like `DecimalDtypeType` for example.

**Review by kkraus14 — APPROVED** [review]

**Pull request merged**